### ISSUE TITLE
Suggest cleanup of geometry handling in the CPU version

### ISSUE THREAD

**Issue by mondracek** [ProkopHapala/ProbeParticleModel#198]
As I mentioned in my comment to #52, the way the CLI version of our code handles the geometry parameters (`params['gridA'], ... params['gridN']`, `lvec`, `nDim` etc.) could use some cleanup. For starters, I propose we discuss this issue:
1. What do we expect to be the role of `gridA`, `gridB`, `gridC`, and `gridN` parameters in the code? Do we want them to be just an optional input (in case the geometry is not specified by other means, in particular in a `*.xsf` or other input file)? In such a case, they can be safely forgotten (I do not mean literally deleted, just ignored) after the initialization stage. The actual lattice vectors and grid size should be consistently carried in variables like `lvec` and `nDim` throughout the actual computation. Alternatively, do we want `parms['gridA'], ... params'['gridN']` to serve as a global variable? Then, we should consistently work with those. In the most extreme version of this choice, we would not even need to introduce `lvec` and `nDim` variables at all. I can somehow understand that a compromise might be desirable: Perhaps we want some functions to be also usable as stand-alone entities, without being called from the main scripts. A moderately experienced user might want to do something in an interactive Python session. The `loadGeometry` function is one of those we may want to be able to call this way. Then, it makes sense for such a function to take `lvec`, 'nDim` etc. as arguments and store them as its local variables. On the other hand, some other functions may be intended as mere subprograms of the main executable scripts and need access to global variables, since passing everything they need through arguments to them would be inefficient and would generate too long argument lists. I suspect this compromise was what @ProkopHapala had in mind when originally writing the code. Prokop, is that so? I am actually in favor of the compromise solution, given that we decide *very carefully* which functions should be the independent entities, taking geometry parameters as their arguments and perhaps returning them as their values, which functions should in contrast access the global variables, and why the chosen option is needed in every case. Because if we are not clear about this, we inevitably end up with the mess we currently have.
2. As a second point, I would like to make sure that the code reads the origin of the unit cell from the `*.xsf` and `*.cube` files every time, stores it and handles accordingly. That is, the code should not rely on the origin being zero. We might even introduce an optional parameter `params['gridOrigin']` to `params.ini`, which would be used with `*.xyz` files. That should solve the problem with negative coordinates of atoms in non-periodic systems, which is currently effectively forbidden.

**Comment by mondracek** [comment]
I failed to reproduce what I thought was the most severe issue that motivated this proposal for a thorough cleanup with regards to how the geometry is being handled. I was under the impression that sometimes, the grid/lattice parameters read from `params.ini` are taken into account even though they should have been overwritten with a `*.cube` input. I do not know what happened then, probably some error in one of my `*.cube` files that disappeared after I had recalculated everything, but now it looks as though the problem was not with the `ppafm` code. Anyway, there are still several open issues related to geometry in one way or another, like the recent #201, #200, #199, the older #190 and #6, or even #52, though I have created a pull request associated with the last one recently. I would still prefer forcing some consistent logic onto the geometry handling along with solving those issues rather then just close those with some *ad hoc* patches that would somehow work until they'd bite us to our backs with any new modification of the code.

**Comment by mondracek** [comment]
I began reviewing the code and now it seems to me that what I will advocate with respect to the geometry will need not be any big revolution after all. I have looked into `core.py` and I think I gather the logic followed there. The primary source of information on lattice vectors is typically a argument called `cell`, an analogy of `lvec` from `HighLevel.py`, `coommon.py` or `io.py`, but the global parameters , `params['gridA']`, `params['gridB']`, `params['gridC']` may be used as a backup source if `cell` is undefined. I believe this is a good way to do things, so I will try to keep this logic consistently.
I have a question about the function `setFF`, defined here: https://github.com/Probe-Particle/ppafm/blob/2344a7b01000bbf8ce3d2ddbf92abf3b9fe917c9/ppafm/core.py#L70
Do we even need it? I mean, it seems to be used only in some obscure examples in `examples/_bak/`. Do we want to have these as a part of the `ppafm` package? If not, I suggest we remove those examples along with the function `setFF`. Otherwise, if we want to keep it, I suggest the following small cleanup: `setFF` seems to reach for the global `grid` parameters twice if `cell` is `None`; first directly in its body and second time (but this never happens because by then `cell` has been already defined by the first case) through `setFF_shape` which in turn calls `setGridCell`. I would remove the test of `cell is None` and the subsequent assignment of `gridA`, `gridB`, `gridC` from `setFF` , relying on this being done in `setGridCell`.
Moreover, if we use `gridA`, `gridB`, and `gridC` as the default for the lattice vectors in `core.py`, I believe we should also use `gridN` as the default for the grid size there (in particular, probably in the `setGridN` function). This enhancement should in fact not change the behavior of our code at this moment, because the flow of the code is now such that the `core.py` functions are always called for some pre-established `FF` array and learn the grid from the shape of this array. That is fine and should remain that way, but I still think the `setGridN` should check the `n` argument for being `None` and use `gridN` as the fallback in such a case, just to make the code more robust.

**Comment by mondracek** [comment]
Just to clarify how I expect the job related to this issue would be done: I will do it, of course, but only after hearing the opinion of others.

**Comment by mondracek** [comment]
Another candidate for deletion: `wrapAtomsCell` in `common.py`. Seems to be used nowhere. The function it implements, as far as I can tell, is this: shift positions of atoms by `[da + 10, db + 10]`, then fold them back to the first unit cell. Might be useful perhaps, but at the very least, if we are going to keep it and use it, the artificial and mysterious shift by `10` should be removed.

**Comment by ProkopHapala** [comment]
@mondracek 

>  I suspect this compromise was what @ProkopHapala had in mind when originally writing the code.

yes, it is true I was aiming for such compromise. But it is also true some structures in the code are just for historical reasons (to minimize effor of rewritining parts of code which are already there)

* In general I think it is reasonable to consider `lvec` and `nDim` to be main storage of information about the lattice. `gridA`, `gridB`, `gridC` are mostly just for user input in `input.ini` (so that each line in the input-file has some name)
* I think some trubles with molecular geometry and lattice whould be reduced if we define class which pack them togheter (including the origin of the lattice). I know people were suggesting to use [ASE](https://wiki.fysik.dtu.dk/ase/) but to me it seems overkill (introduce dependnece on huge 3rd party package into core of ppafm - that I don't like). I would rather make something like [class AtomicSystem](https://github.com/ProkopHapala/FireCore/blob/27547b6dad744bf79641edbf4bbcb5c2cd95f1f6/pyBall/atomicUtils.py#L881) which I have in my other project (it encapsutlates reading and loading, keeps the lattice vectros, originn, atomic types, positions, parameters, etc.). Than we can pass instances of that class into fuctions (rather than passing arrays indepenendtly), which simplyfy the interface. 
   * but I'm not sure if to do it now, since it is quite large reoganization of the project 

>  I would still prefer forcing some consistent logic onto the geometry handling along with solving those issues rather then just close those with some ad hoc patches that would somehow work until they'd bite us to our backs with any new modification of the code.

I'm very much pro. If you have ideas pleas go ahead with proposing them. I'm very open to it. I'm just not enough systematic myself, and also currently it is not my priority.

ad `setFF`

what you meanit is not used? OK, maybe `setFF()` is not used, but `setFF_shape()` and `setFF_pointer()` are used a lot. Without them you cannot share forcefield between python and C++ part.

So yes, this is also perhaps more historical. Firts I wrote `setFF()`, then I realized that often I need more fine-controle (to set shape and data of the grid forcefield independnetly). So I factored out  `setFF_shape()` and `setFF_pointer()`. So yes, perhaps now `setFF()` is almost not used.

ad `wrapAtomsCell` - I don't really remember what I used it for. Perhaps I used it just once for some debugging.

**Comment by mondracek** [comment]
@ProkopHapala, thanks for the reply. As for the `setFF()` function, yeah, I mean `setFF_shape()` and `setFF_Fpointer()` or `setFF_Epointer()` are called directly whenever one might want to call `setFF()`. So I suggest we either delete the definition of `setFF()` or we actually use it where appropriate. Perhaps we can unite `setFF()` and `prepareArrays()` into one function and use that one. Now, as I understand it, both functions are doing similar jobs but
  1. `prepareArrays()` does not call `setFF_shape()`. We always call `setFF_shape()` shortly after `prepareArrays()`, however
  2. `setFF()` calls `setFF_shape()` but, in contrast to `prepareArrays()`, it cannot create a new array if the corresponding aray argument is `None`

**Comment by mondracek** [comment]
Now I see that the `C` functions `setGridN` and `setGridCell` are defined in both `ppafm/cpp/GridUtils.cpp` and `ppafm/cpp/ProbeParticle.cpp`. The definitions are identical so the duplicity should not cause any problem until we happen to change one of them, but still, it does not feel correct.

**Comment by ProkopHapala** [comment]
> Now I see that the `C` functions `setGridN` and `setGridCell` are defined in both `ppafm/cpp/GridUtils.cpp` and `ppafm/cpp/ProbeParticle.cpp`. The definitions are identical so the duplicity should not cause any problem until we happen to change one of them, but still, it does not feel correct.

I agree that it is not nice from the point of code structure (modularity etc.). But the point is that `GridUtils` and `ProbeParticle` both compiles into independnet binary (.so) and they does not depend on each other.

But yes, it would makes sense to factor out these duplicite things into singe header file (.h). But I'm not sure how it works if the function is in `extern C{}` (if it is not problem to have `extern {` in header file ?)

**Comment by ProkopHapala** [comment]
> 1. `prepareArrays()` does not call `setFF_shape()`. We always call `setFF_shape()` shortly after `prepareArrays()`, however
> 2. `setFF()` calls `setFF_shape()` but, in contrast to `prepareArrays()`, it cannot create a new array if the corresponding aray argument is `None`

OK, to be honest I don't remember exactly the code structure and the rational nor the historical path which lead to this. If you find it nicer that way, feel free to change it as you say

**Comment by mondracek** [comment]
Good. Concerning the duplicity of `setGridN()` and `setGridCell()` function, I'd rather leave it as it is for now because I am also not sure how exactly the `C` linking works and I might actually break it by trying to repair it. I guess moving the duplicate into a header *should* work in principle, although the header files might not be exactly intended for such purpose, because including a header file into a `*.cpp` file through `#include` should be (as long as I understand it) done by the preprocessor essentially by pasting the plain text of the `*.h` file into the `*cpp` file before the compilation. But I am not quite sure. If anyone of you other guys knows how to handle this in `C` properly, you are welcome to propose a solution.
Otherwise, I see now that the cleanup I wanted to make will not require as radical changes of the code structure as I thought when I started this issue. So I will rather pick up more specific tasks concerning the handling of geometry (and force field grids) and I will do the cleaning along the way.
One of those specific issues I am planing to address would be that we introduce the proper handling of the origin as recorded in `*.xsf` and `*.cube` files. See the second bulleted point in my first comment in this thread. Do you agree that this should be done? Other such issues are those referenced in my second comment here (#6, #190, #199, #200, #52) .

**Comment by ProkopHapala** [comment]
> I'm also very much in support for packing things into a class object. And I already did something in this direction in the OCL API with the DataGrid (except without the atoms), It could be nice to somehow interface this with the CPU part, maybe make the DataGrid a subclass of the whatever class we create for the core code

Yes, I like your `DataGrid` class. Maybe we should discuss more what ideas you have about class hierarchy or relations? You mention CPU vs GPU. But I would perhpas first speak about geometry (i.e. atomc coordinates and types) vs grid (3d array) vs lattice vectors. 

1. One option is to make one dynamic class, which have optional fields (e.g. initialized to None). e.g. it may or may not have data grid, it may or may not have atoms, or lattice vectors.
2. Another option would be to make hierarchy. E.g. starting from `Atoms` class, then add lattice vectors `AtomsPBC` and then ad grid (array) ... or in other order?
3. Another option is to make 2-3 independnet classes e.g. (i) `Atoms` which strores atomic positions and types (ii) `Grid` which stores lattice vectors and optional data array. And them pack them together in e.g. `AtomGrid` class

Because python in very dynamic unlike C++, It woudl maybe prefer (1) i.e. to have optional dynamic fields. The disadvantage is that it would require to put everywhere checks if some field is initialized, like `if( lvec is None  ):` 

> Also, I already used ASE as an optional dependency in the GUI for the geometry viewer.

Don't understand me wrong. I'm very much pro integration with ASE as **optional** dependency. I also use ASE and I like it. But I don't like the idea to depend on it in such core functionality as loading and manegind atomic coordinates (i.e. not being able to run ppafm at all without ASE)

**Comment by NikoOinonen** [comment]
> 1. One option is to make one dynamic class, which have optional fields (e.g. initialized to None). e.g. it may or may not have data grid, it may or may not have atoms, or lattice vectors.
> 2. Another option would be to make hierarchy. E.g. starting from `Atoms` class, then add lattice vectors `AtomsPBC` and then ad grid (array) ... or in other order?
> 3. Another option is to make 2-3 independnet classes e.g. (i) `Atoms` which strores atomic positions and types (ii) `Grid` which stores lattice vectors and optional data array. And them pack them together in e.g. `AtomGrid` class

I would avoid deep class hierarchies (2). They become very rigid and difficult to change when new needs arise, and it's often not clear at which level of the hierarchy some piece of data should exist.

So I would opt for [composition](https://en.wikipedia.org/wiki/Composition_over_inheritance) instead (1, 3). I see 1 and 3 being structurally the same. E.g. either we store the lvec and grid as `np.array`s or some custom classes of our own, but in any case they exist as properties of the storing class, and that structure can be changed internally flexibly.

> I have to say, that one thing I hate about python is that due to dynamic initialization of class properties, it is often not clear what properties the class is supposed to have when reading the code (at least comparing to C++). This was actually main reason why I considered classes in python poinless for quite long time. Do you know some good practicies how to make the declaration of class properties in python more explicit?

One way to do this is to make all of the internal variables private (prefix with underscore `_`) and then make an interface by using the `@property` decorator for class methods. For example, in the `DataGrid` I store the host numpy array like this: https://github.com/Probe-Particle/ppafm/blob/4e04d47cefac340ab734a2ef4000cc901793c5b0/ppafm/ocl/field.py#L118
but it can also sometimes be `None` because it exists on the GPU instead. And then I create a property method that returns the array after checking that it's fine: https://github.com/Probe-Particle/ppafm/blob/4e04d47cefac340ab734a2ef4000cc901793c5b0/ppafm/ocl/field.py#L152-L158

The `@property` decorator makes it so that you can access the method as if it's just an attribute, e.g.
```python
array = datagrid_instance.array
```

With the doc string in the method, it also makes for nice documentation, so the API is clear: https://ppafm.readthedocs.io/en/latest/ppafm.ocl.html#ppafm.ocl.field.DataGrid.array

**Comment by NikoOinonen** [comment]
> I'm not sure which is best approach. Because python in very dynamic unlike C++, It woudl maybe slightly prefer (1) i.e. to have optional dynamic fields. The disadvantage is that it would require to put everywhere checks if some field is initialized, like if( lvec is None  ):

We can probably make this simpler for ourselves by insisting that some of the fields always exist. The `lvec` we can for example read from the default parameters if it's not set, as Martin suggested above.

I think this consideration also encourages splitting the functionality into multiple classes (option 3.), because then if you only need the grid, for example, then you can just use the `Grid` class instead of the `AtomGrid`.

**Comment by mondracek** [comment]
* I noticed that the use of classes for storing geometry has been already suggested by @ProkopHapala before, as issue #97. Perhaps we can close that issue noting that the discussion has been transferred here.
 * I have been recently trying to enable reading data from XSF files created by the FHI-AIMS code. There were some issues specific to FHI-AIMS that prevented loading them with the `loadXSFGeom` and `loadXSF` functions from `ppafm/io.py` as these function stand now. I will explain more when I start a dedicated issue about that. What I want to say for now: Working with the XSF files, I noticed that the vectors that define the grid can be in general something completely different from the lattice vectors that define the periodicity. In our code, we essentially work with the grid vectors and when we happen to need the lattice vectors, we pretend they are the same. I am not saying we need to change this approach right now, but we should keep in mind that we might need to differentiate between the *grid vectors* and *lattice vectors* in the future. So when rewriting the code, introducing new classes for the geometry etc., we should do it in a way that at least does not make future splitting of *lattice vectors* from *grid vectors* more difficult than necessary.

**Comment by ProkopHapala** [comment]
> we pretend they are the same. I am not saying we need to change this approach right now, but we should keep in mind that we might need to differentiate between the _grid vectors_ and _lattice vectors_ in the future. 

Not sure I understand. 
* Do you mean that there are situations (systems) where it would be usefull to distinguies the two ?
* Or that in the code there is no checking mechanism which makes sure grid vectros and lattice vectros are the same (while they should be) ?

**Comment by mondracek** [comment]
@ProkopHapala: Both, sort of.
* There may be cases, admittedly not particularly common, in which the distinction might be useful. For example, @NikoOinen mentioned to me when reviewing my PR #216 that the user may be interested in a relatively small scanning region that does not cover the whole system. Then, it is enough if the grid includes just the scanning region plus some padding. This will more likely happen for large non-periodic systems but such scenario is not entirely excluded for periodic systems either. If atom-centered potentials like LJ are going to be used to calculate the force field and if the requested scanning region is situated at or near the periodic cell boundary, the lattice vectors of the true periodicity cannot be ignored even though the grid size should be smaller than the unit cell.
* More often, the grid should cover exactly one periodic unit cell. For example, we rely on this when the scanning region crosses the unit cell boundary and the tip position rolls from one unit cell to another. But we do not check whether the grid axes defined in XSF files really agree with the lattice vectors in the same file, even if we expect that they do.
* In files like `OutFz.xsf`, the grid axis in fact *should not* be the same as the lattice vectors. The grid axes determine the scanning region there and that is usually different from the system's periodic unit cell. At this moment, we simply do not write the lattice vectors into `OutFz.xsf` at all. That does not matter as long as we consider these `OutF?.xsf` files just a temporary storage of intermediate data produced by `ppafm-relaxed-scan` before they are passed to `ppafm-plot-results`. The `ppafm-plot-results` script does not need the information on lattice vectors. However, I do not like this approach in which the files like `Outfz.xsf` are seen as something internal to the `ppafm` code only. Why bother splitting the code into several scripts unless the output from each of the script can stand on its own? Once the user wishes to view and analyse this force fields separately, it would be desirable that the XSF files include information on the position of atoms and on the lattice vectors. We have in fact done the same for the `FF*.xsf` files already.

By now, I have started working on the problem of properly defining the grid origin, as needed to solve #223. On that occasion, I noticed there are parameters called `FFgrid0`, `FFgridA`, `FFgridB`, and `FFgridC`, the purpose of which is exactly to define grid geometry different from periodic lattice vectors, but only for the OCL version (see the discussion under #139). I do not really understand how the OCL version works, but can we perhaps adapt the solution used there for the CLI version too?

**Comment by ProkopHapala** [comment]
@ProkopHapala 

>> the user may be interested in a relatively small scanning region that does not cover the whole system. 

I see, make sens

>> Why bother splitting the code into several scripts unless the output from each of the script can stand on its own? Once the user wishes to view and analyse this force fields separately, it would be desirable that the XSF files include information on the position of atoms and on the lattice vectors.

I agree, I was also considering .xsf files as intermediate result which allows easy debugging and eventually some way to analyze the results. ... Maybe I just don't understand the proper format of .xsf files. I did not saw any problem there. If there is such problem it make sense to correct it.

**Comment by mondracek** [comment]
@ProkopHapala: As for the problems with the XSF format:
* The missing atoms and lattice vectors are more an inconvenience rather than an error. Well, actually, some XSF viewers do require the atoms to be present; we have a placeholder hydrogen atom in the XSF file for that purpose. Would be better to have the real atomic structure of the sample there.
* More closer to being a real bug is the blind adding of extra points on the boundary when saving the XSF file in our code and removing them when reading these files. This is okay internally for our code as it always adds those files when saving and always removes them when reading, so everything compensates. When reading XSF files generated by an external code, those are potentials or electron densities, typically generated by VASP for a periodic system. In such cases, the extra points are really present too, so it is okay that they get removed. But in `OutFz.xsf` and the like, they are usually nonsense.

Let me explain here how (to my understanding) the logic with the extra points on the grid cell boundary is supposed to work in the XSF files. First, if the length of one the grid vectors is _a_=(_a_<sub>x</sub>,  _a_<sub>y</sub>, _a_<sub>z</sub>) and there are _N<sub>a</sub>_ points along that vector, the elementary grid step along that vector is _da_=(_a_<sub>x</sub>,  _a_<sub>y</sub>, _a_<sub>z</sub>)/(_N<sub>a</sub>_-1). This is always true. It is part of the definition of the XSF format and it is non-obvious, you can imagine a different format that would have _N<sub>a</sub>_ instead of _N<sub>a</sub>_-1 in the denominator of that definition. For a general grid the dimensions of which have nothing to do with the periodic unit cell, this is end of the story. However, there is a little problem when we want the grid to cover exactly the unit cell. Imagine that _a_=(_a_<sub>x</sub>,  _a_<sub>y</sub>, _a_<sub>z</sub>) is now the lattice vector. We want to sample the cell by _N<sub>a</sub>_ points along this vector with the grid points being (of course) equally spaced and, for starters, they are supposed to be unique, they do not repeat themselves because of the periodicity. The spacing between the points should of course be _da_=(_a_<sub>x</sub>,  _a_<sub>y</sub>, _a_<sub>z</sub>)/_N<sub>a</sub>_ under this condition. So what should the corresponding *grid vector* be? Right, considering the formula that defines the relation between the grid vector and the grid spacing, it has to be _a'_ = _a_*(_N<sub>a</sub>_-1)/_N<sub>a</sub>_

But this is inelegant, we would prefer to have _a'_ = _a_ and this is what most codes (but not FHI-AIMS!) generating XSF files for periodic structures chose to enforce. The price to pay is adding a last point along every axis that is a identical copy of the first point by the lattice periodicity. So now, we have _N<sub>a</sub>_ +1 points instead of _N<sub>a</sub>_ and the last point is kind of superfluous in terms of information content, but the lattice vectors and grid vectors are the same, which makes us happy, because we can immediately see (looking at the XSF file header) that the grid indeed covers the whole unit cell. So, to some up
* When writing the XSF files, the extra points should be added only if the grid covers exactly the entire periodic unit cell, otherwise, they make no sense and are actually wrong (because why should the last point be identical to the first one when you extend a grid that does not fit the periodicity of the structure?).
* Similarly, when reading the XSF files, the the end points should be discarded as the extra points only if the grid dimensions coincide with the lattice vectors. Now, the lattice vectors should be read right from the XSF file and compared to the grid vectors read from the same file (the XSF file contains both, if properly written, unlike the CUBE file).

**Comment by mondracek** [comment]
@ProkopHapala Is there any good reason why the order of indices for the grid array is FF[iz,iy,ix] rather than FF[ix,iy,iz] in the CLI version? I used to think this is needed for the CPP parts of the code to be efficient but now as I look deeper into it, that does not seem to be the case.

**Comment by ProkopHapala** [comment]
>  Is there any good reason why the order of indices for the grid array is FF[iz,iy,ix] rather than FF[ix,iy,iz] in the CLI version? I used to think this is needed for the CPP parts of the code to be efficient but now as I look deeper into it, that does not seem to be the case.

I think efficiency-wise it is rather opposite (we compute each z-stroke independently => better if z is the fastest axes in the array )

I think the reason for FF[iz,iy,ix] ordering was perhaps plotting ? (z-slices), but it is not very good motivation (matplotlib can plot FF[:,:,iz] as well as FF[iz,:,:] ). Maybe it is somehow useful in Fz->df conversion (but perhaps not), or for saving the data to .xsf ?  I don't know. Eventually we may change the order  ... but it would require changes on many places in the code ... co unless there is good reason .... ?

But

**Comment by mondracek** [comment]
@ProkopHapala, thanks for the explanation with respect to the array ordering. As for the Fz->df conversion and saving `*.xsf` files, I happen to know something about that stuff and no, the reversed index order is not needed for it. In the `Fz` to `df` conversion, it does not matter, the conversion is in the end just a convolution of `Fz` with the appropriate kernel along the `z` axis and we would literally only have to change `axis=0` to `axis=2` as an argument to an `np.apply_along_axis` function at one place in the code in order to adapt the `Fz2df` function to the natural ordering. In reading and saving `*.xsf` files, the reversed ordering actually complicates stuff because there is a difference between the CLI and OCL version in the array ordering so the `loadXSF` and `loadCUBE` functions need a `xyz_order` argument which tells them whether they should transpose the array or not.  Now I see that `saveXSF` does not distinguish between CLI and OCL versions but rather always assumes the reverse order. I did not notice this last fact before. I do not know how it works in the OCL version, perhaps the ordering is really not consistent, arrays which we need to read to XSF are ordered one way, arrays we need to save to XSF the other way and we hope there is nothing we want to read, modify without doing the transposition and then save. I was about to offer to volunteer in changing the ordering so that it would always be FF[ix,iy,iz] but I thought I would not require touching the OCL part. Now that I see the issue affects both CLI and OCL versions, I am not sure I want to do it.

**Comment by NikoOinonen** [comment]
@mondracek As far as I understood, the zyx ordering in the IO functions was mostly because that is how the XSF file format is defined. I specifically added those `xyz_order` arguments to the functions so that I could get the arrays in the xyz order to use in the OCL code.

It should not be a big change. I think the only places in the OCL code that do IO related things are these two methods: https://github.com/Probe-Particle/ppafm/blob/9f009aec25d8d895f4168f061501211a37a9dd1c/ppafm/ocl/field.py#L192-L195
https://github.com/Probe-Particle/ppafm/blob/9f009aec25d8d895f4168f061501211a37a9dd1c/ppafm/ocl/field.py#L227-L230
The `to_file` method in particular transposes the array before saving.

**Comment by mondracek** [comment]
@NikoOinonen, thanks for the explanation. So @ProkopHapala was right after all that the likely original motivation of the zyx order in the CLI code was reading and saving XSF files. The native ordering in the XSF file is the opposite of what we'd need, either in CLI or OCL. Seems that historically, the approach in CLI was to accept that ordering and interpret the array indices as ZYX while `transpose` was performed for OCL. Okay, so we would not have to touch the OCL version at all in order to introduce consistent usage of the XYZ order but thorough changes would be needed in CLI. When done, the I/O functions would not have to differentiate between CLI and OCL and the CLI code would be a bit easier to understand but there is a big danger of introducing errors into the CLI code while implementing the change. Though the errors should be easy to spot by testing, a wrong data order on the grid would destroy the resulting figures completely. What do you think, is it worth the effort? I will not push the change at this moment myself but if anyone else would like it to be done, I am still ready to implement it.

**Comment by mondracek** [comment]
As for the `OutFz.xsf` files I mentioned earlier in this tread, I see that the inclusion of the sample atomic structure in those files has been already implemented. I was probably looking art files produced by an old version of the code when writing my earlier comments. But the extra points at the scanning region boundary are probably still there.

**Comment by ProkopHapala** [comment]
>> What do you think, is it worth the effort? 

I think it is woth it.

>> I will not push the change at this moment myself but if anyone else would like it to be done, I am still ready to implement it.

if you implement it, I'm pro :-) 
I think it is worth it, but perhaps does not have hi priority.

>> Though the errors should be easy to spot by testing, 

Copy that

**Comment by mondracek** [comment]
So, I am going to change array indexing in the CLI version from `FF[iz,iy,ix]` to `FF[ix,iy,iz]` as discussed above. Now. The reason is:
* I am working on #337 (which can be considered another sub-issue of this broad issue, by the way).
* In order to solve #337, I have written a new `loadXSF` function and then, I realized my new version of `loadXSF` was not working with `ppafm`, because it did not reverse the order of indices as expected by `ppafm`.
* I do not want to spoil my nice `loadXSF` function by adding this extra reordering of indices there. Moreover, in the process of solving #342, we would have to deal with the order of indices too.
* So in the end, though the change is not strictly necessary, I want to consider it a prerequisite for #337 and #342 and I am going to implement it .

**Comment by NikoOinonen** [comment]
> Uff, but if you change order of indexes, it has far reaching consequences on the C++ kernels which have to be completely rewritten as they assume some indexing. I think it is much easier to transpose arrays on I/O (loading and saving to files like .xsf, .cub etc.) then reordering internal kernels.

The indexing in the C++ code is dependent on the memory layout, which is not quite the same thing as the order of the indices in the Python code, e.g. numpy has an option to layout the memory in C order or Fortran order, but it does not affect how you index or slice the arrays in Python. So, I'm guessing this does not necessarily have to affect the C++ code.

**Comment by mondracek** [comment]
Hi, @ProkopHapala , yes, I realized soon after I wrote my last comment that what I want to do would also require changing the C++ part. @NikoOinonen , while technically you are right that the C++ code depends on the memory layout, in the end of the day, assuming C ordering for the `numpy` arrays, the `[iz,iy,ix]` order of indices in Python means `ix` is going to be the fastest-changing index when iterating over the array in the order in which it is stored in memory, while `[ix,iy,iz]` indexing in Python means `iz` is going to be the fastest-changing index. Trying to change this simply by going from C ordering to Fortran ordering with the `numpy` arrays would hit a problem for vector fields. Those are currently indexed as `[iz,iy,ix,i]` in Python, where `i` determines the vector component. Switching to Fortran ordering of `numpy` arrays (without adapting the C++ code) would make us use not `[ix,ty,iz,i]` indexing, as probably intended, but rather `[i,ix,iy,iz]`: The vector index would have to come as the first index (because we definitely want all three vector components to be stored right next to each other in memory). 

Anyway, I decided I would still go for unifying the index order under *some* logic, even if I have to rewrite parts of the C++ code too. I briefly considered giving up and leaving the indices as they are now but then I realized that, as we have also already mentioned in the above discussion, the currently used ordering differs between the CLI and OCL versions. @NikoOinonen , is it so? If yes, that kind of trips the balance for me towards the decision to remake things into a more consistent state.

So, do we prefer an index order of `[ix,iy,iz,i]` or `[i,ix,iy,iz]` for the arrays that represent vector fields? @ProkopHapala , @yakutovicha , @NikoOinonen , can you please vote on that?

**Comment by NikoOinonen** [comment]
@mondracek I'm not 100% sure, but it looks to me like the force field components would have to be the last index in C order, because the force field components in the C++ are packed into `Vec3d` objects which are defined as a class: https://github.com/Probe-Particle/ppafm/blob/c6e7e7416e398178ca74c6aec1cc978d655c5405/ppafm/cpp/Vec3.h#L17-L27
The force field is an array of these objects. The fields in the object probably have to be contiguous in memory.

**Comment by NikoOinonen** [comment]
And for the OCL version, the indexing on the Python side is in `[ix,iy,iz,i]` order, but the memory layouts are a bit messy. It's mostly using C order, but there are some places where we use OpenCL built-in functions which are in Fortran order so we switch it up. There are some wrappers in the Python code that fix the order when we return to numpy arrays and the indexing is not consistent, e.g.: https://github.com/Probe-Particle/ppafm/blob/c6e7e7416e398178ca74c6aec1cc978d655c5405/ppafm/ocl/field.py#L1071-L1074 https://github.com/Probe-Particle/ppafm/blob/c6e7e7416e398178ca74c6aec1cc978d655c5405/ppafm/ocl/field.py#L1086-L1087

**Comment by mondracek** [comment]
Looking at the piece of code @NikoOinonen is showing from `ppafm/ocl/field.py` it seems that the array needs to be physically reordered indeed. But if that happens only once for the whole field when the field is loaded and then for the second time reordering the field back when it is going to be saved, maybe it is not that bad. However, if we (contrary to my vote in my previous comment) settled for `[i,ix,iy,iz]` indexing and Fortran ordering, this reordering inside `field.py` might not be necessary, right?

**Comment by NikoOinonen** [comment]
> Also, when the OCL code changes the memory ordering, does it literally reorder the numbers in memory? That would be a huge inefficiency, wouldn't it? Or does it just switch the order of indices along with the switch between C/Fortran ordering so that everything remains physically the same in the memory?

The memory layout is changed only once as a part of another operation, so it does not cost that much, but probably a bit inefficient, yes.

> However, if we (contrary to my vote in my previous comment) settled for [i,ix,iy,iz] indexing and Fortran ordering, this reordering inside field.py might not be necessary, right?

That would probably be the most consistent option. The OpenCL operations do require Fortran order in some parts, so it would actually be the most efficient if we can agree to use Fortran order everywhere. However, it would require rewriting some of the OpenCL kernels and changing the convention for the indexing, which would probably break some user's scripts using the Python API.

**Comment by mondracek** [comment]
Thanks for the explanation. I am for the **`[i,ix,iy,iz]`** index ordering with **Fortran** memory layout then. I might even be lucky not to break the C++ functions with that, because when C++ interprets a data array with this kind of memory arrangement as if it were in C layout, it is going to correspond to `[iz,iy,ix,i]` indexing and that is what we currently assume in the C++ parts anyway. I need to think this over a little bit but now it really seems to me that the C++ part will be fine without any changes. 
By the way, the API. Do we really know about any users out there using their own scripts relying on our API? Because my motivation for pushing these changes, although they do not directly affect the functionality or introduce new features into `ppafm`, is that whenever I was trying to use the `ppafm` as a collection of functions and modules to be interfaced with my own scripts, I have always failed, getting completely lost in the mess of inconsistencies like what exactly the grid parameters meant and what the shape of data array was supposed to be in each particular part of the code.

**Comment by ondrejkrejci** [comment]
@mondracek and @NikoOinonen - from my point of view I do not see any anti-reason why not to go with it. Classical PPSTM is using only the output positions of PPAFM, thus if the inner transformation is done properly, it is not affecting PPSTM as for the STM depending on the OCL Python API, I am not quite sure, but I still think, that this should not be directly connected for the inner grids, but I can be wrong, as I have not seen the code. @NikoOinonen ask then. We can even have a meeting about this quickly, to move things forward.

Otherwise, consider having green traffic-light from me.

**Comment by mondracek** [comment]
@NikoOinonen, I had to look more carefully into the code of the OCL version while trying to implement the index ordering and I see I am not able to change it myself. I do not understand the structure of the code well enough. So I will try not to touch it except for the places where it interfaces with functions that are shared by both OCL and CLI versions and where I would change those functions. Anyway, I would still like to make my final decision about which ordering to use with the OCL in mind, just in case we wanted to unify the index ordering and memory layout in that version too, perhaps later. So currently, two types of ordering are being used in OCL: What I call **Fortran[i,ix,iy,iz]**, which is required by the OCL buit-in functions, and **C[ix,iy,iz,i]**, which is required by some of our OCL kernel functions in `ppafm/ocl/cl/`, while others of the `ppafm/ocl/cl/` still work with the **Fortran[i,ix,iy,iz]** ordering. The arrays therefore need to be reordered inside OCL and that cannot be easily changed, not without rewriting the `ppafm/ocl/cl/*.cl` kernel functions. By the way, I wonder, what is the built-in OCL function which requires the Fortran memory layout? I could not identify this in the code, I am unfortunately getting lost here.
To recap, the candidate options for the unified ordering, at least for the CLI version so far, are the two following:
* **Fortran[i,ix,iy,iz]** This would most likely not require rewriting the C++ (`ppafm/cpp/`) part of the code. Moreover, if we want to unify the OCL too at some point and if it is indeed true that the OCL built-in functions strictly require the Fortran memory layout (but see a footnote below), this might be the only option. However, we would need to distinguish the vector component in force fields by the first rather than last index in Python parts of the code, which is different from the notation we are using now.
* **C[ix,iy,iz,i]** This would require rewriting the C++ part, but I can do it. The vector component of fields would remain to be the last index in Python. The `iz` index would become the fast-changing one when iterating over the arrays in C++, which may parhaps make somewhat more sense when simulating tip oscillations along the *z* axis, otherwise, it does not matter.

<ins>Footnote</ins>: Why would the Fortran ordering be required? Are these built-in OCL functions not Python-agnostic? If they are, the `[i,ix,iy,iz]` order in Fortran layout is the same as `[iz,iy,ix,i]` in C layout, so now, it is just about interchanging the `ix` and `iz` index, but why would the fuction need to know which axis is the *z* axis?

**Comment by mondracek** [comment]
In the above summary, I forgot to add that the **Fortran** order is used to store data in the **XSF** format, the format we use as the default in `ppafm`. That might have been the original motivation for the reverse order of indices: `C[iz,iy,ix]` corresponds to `Fortran[ix,iy,iz]`. However, as I understand the present workings of our code, the functions like `loadXSF` do not actually take any advantage of this, they are transposing the arrays back and forth anyway.

**Comment by mondracek** [comment]
One more argument for **C[ix,iy,iz,i]**: C memory layout is the default for multidimensional python arrays. If we pick the Fortran convention, we need to explicitly specify Fortran ordering every time we create or copy an array, so as to make sure we preserve the layout.
Anyway, now I lean towards taking the OCL bulit-in functions as a tie-breaker: So again, @NikoOinonen , what are those functions?

**Comment by NikoOinonen** [comment]
>  By the way, I wonder, what is the built-in OCL function which requires the Fortran memory layout? I could not identify this in the code, I am unfortunately getting lost here.

We use the `read_imagef` function for interpolating the force field during the probe particle relaxation process: https://github.com/Probe-Particle/ppafm/blob/4097baffd749a260e518da9955c2ab5cfe9b4d92/ppafm/ocl/cl/relax.cl#L72

> Why would the Fortran ordering be required? Are these built-in OCL functions not Python-agnostic? If they are, the [i,ix,iy,iz] order in Fortran layout is the same as [iz,iy,ix,i] in C layout, so now, it is just about interchanging the ix and iz index, but why would the fuction need to know which axis is the z axis?

I'm not sure exactly what we are talking about, but when we are computing distances in 3D space we definitely need to know how the axes are ordered when we convert the linear indices into 3D indices like so: https://github.com/Probe-Particle/ppafm/blob/4097baffd749a260e518da9955c2ab5cfe9b4d92/ppafm/ocl/cl/FF.cl#L521-L524
We also need to convert between the OCL buffers and numpy arrays, keeping the ordering of the data consistent in between. In numpy we get two options, either the first index changes fastest, or the last index does. Of course we could write wrappers etc. to order the indices however we want, but that adds complications.

**Comment by NikoOinonen** [comment]
Actually, @ProkopHapala do you remember if there is any other reason we use the OCL `image3d_t` arrays instead of buffers than for the interpolation during the relaxation? I see that we have implemented the trilinear interpolation also manually (actually twice), so we could easily switch to using the buffers where we can read/write them in any order we like. Maybe there is some performance implication?

**Comment by mondracek** [comment]
@NikoOinonen, I see, thanks. What I was trying to say in the unclear part of my comment related to the OCL functions was that the function (`read_imagef`) only cares about the physical ordering of data in memory and does not care whether we think about that particular memory layout as Fortran layout with the corresponding order of Python indices or C layout with the opposite order of Python indices. What I did not know was that the function uses a 
3D (or actually 4-dimensional?) indexing of its own with the index passed an argument of the `float4` type, so the relation between these `float4`-type index and the position in memory is what defines the required memory layout. Okay, so, on the one hand, I do understand now why a particular memory layout is needed for OCL, on the other hand, I think if we had a field in a "wrong" memory layout, we could solve it just by reshuffling the indices in places like
https://github.com/Probe-Particle/ppafm/blob/4097baffd749a260e518da9955c2ab5cfe9b4d92/ppafm/ocl/cl/relax.cl#L53-L60
instead of having to reshuffle the whole field in the memory. In any case, as I wrote above, I am going to leave the OCL part alone for now; I will only unify the array indexing and memory layout for the CLI part.

**Comment by mondracek** [comment]
All (@NikoOinonen, @ProkopHapala, @ondrejkrejci, @yakutovicha, @aureliojgc ...), I have finally decided to go with the **C[ix,iy,iz,i]** ordering in CLI. Any protests?

**Comment by ProkopHapala** [comment]
Martin, C[ix,it,iz] is probably the best performance wise (although the gaing will be probably small). This is my guess I don't have in mind all the cases where it might be more convenient use different order, but I guess the performance critical parts in C\C++ can benefit from this change. In Python I think it does not matter to much what the order is (one only need to use it consistently)

But I'm not sure if the work required to change the array order in whole code is worth it.

I don't see clearly your motivation. I guess it is some sort of elegance and consistency. Maybe it will help to maintain the code in future, but the investment of your time seems to me rather large and definitely it would not be high on my priority list.

**Comment by ProkopHapala** [comment]
> for the interpolation during the relaxation? I see that we have implemented the trilinear interpolation also manually (actually twice), so we could easily switch to using the buffers where we can read/write them in any order we like. Maybe there is some performance implication?

I think textures should be faster than buffers even if hardware interpolator is not used. This should be because of spacial-cashing which understand collocation of voxels in 3d
https://www.perplexity.ai/search/on-nvidia-gpu-is-i-interpolate-MfwKfvK8RkCHVuotbM9EMw
But I did not tested it. In FireCore I implemented tricubic B-spline interpolation using buffers (because it was more strighforward to convert it from C/C++ code) and now I'm tinking to use texture instead of buffer for performance reasons.

**Comment by ProkopHapala** [comment]
> One more argument for C[ix,iy,iz,i]: C memory layout is the default for multidimensional python arrays. If we pick the Fortran convention, we need to explicitly specify Fortran ordering every time we create or copy an array, so as to make sure we preserve the layout.

Not sure I share your mental model. I never understood what is meant by that C-vs-fortran convention. If the first index is fast moving (Fortran?) or the last index is fast moving (C) ? Anyway, I always think in C-convention (I never like fortran and I was only forced to it by fireball). But still one can ask legit question - should the fast axis by x or z ? If we think in images in x,y plane it is natural think about 3D texture as a list of x,y-slices (this was my mental mondel when I started writing probe-particle code). Neverthelss, if one iterate most often over z-direction (e.g. doning z-approach or df-convolution), than it is actually more natural to think about the 3D texture as a box of z-lines.

I don't really care how data are handled in python (I consider python just as a script around C++ or OpenCL core). The very reason why I like using python is that I can bastract from this array packing, and I can use easy slicing (array views). For me is important how convenient is calculation of voxel address inside C or OpenCL (which is mapping from 3d to 1d buffer), and how fast is CPU/GPU cache.

**Comment by ProkopHapala** [comment]
> The arrays therefore need to be reordered inside OCL and that cannot be easily changed, not without rewriting the ppafm/ocl/cl/*.cl kernel functions. By the way, I wonder, what is the built-in OCL function which requires the Fortran memory layout?

I don't think there are any "buildin" functions. (except FFT, but that works along all axis the same way I think). There are only kernels written by us which operate on indexes, is it is our responsibility to choose convenient indexing for the job.

**Comment by ProkopHapala** [comment]
The textures are used when reading (interpolating) buffer when writing (projecting). I don't think for writing there will be any meaningful performance benefits because there must be synchronization (no cache), writing is always slow, but we do hundreds atoms before we write one Voxel so that writing cost is amortized 

Ok, if you think Niko also think it is a mess, and it is worth the effort then let's do it. But I think it is quite big change which will have many consequences. I woul not spend time on that if it is on me.

Performance wise I think the memory layout where z is the fastest axis is perhaps the best as we most often iterate over z-axis.

**Comment by NikoOinonen** [comment]
I can make whatever changes needed to the OpenCL side of things. I don't actually think it's that big of a change, and we actually have a decent number of tests for the code so refactoring should not be too hard.

@mondracek Just make whatever changes you need, and I'll follow up on it.

**Comment by ProkopHapala** [comment]
Hi, not sure how much important it is for this dicussion. But yersterday I was benchmarking speed of tricubic B-spline interpolation implemented using texture vs buffers.

* To my surprise it turns out that the buffer-based solution is roughtly 2x faster than texture based using  `read_imagef`
    *  maybe we some other sumpler settings it could be different, but I was not able to make it work othrwise (currently I have `sampler_t sampler_norm_rep_near = CLK_NORMALIZED_COORDS_TRUE | CLK_ADDRESS_REPEAT | CLK_FILTER_NEAREST` ) 

The in script [test_sample_grid.py](https://github.com/ProkopHapala/FireCore/blob/prokop/tests/tCUDA/test_sample_grid.py)  I tested $10^6$ independent particles move in GridFF potential for $10^4$ steps of dynamics. The resulting runtimes were like: 0.89 seconds with buffer based solution ( see [inline float4 fe3d_pbc_comb()](https://github.com/ProkopHapala/FireCore/blob/f3423ab01e94a9021a7175250ac8f8f9f7005d5f/cpp/common_resources/cl/relax_multi_mini.cl#L1200) and [__kernel void sampleGrid()](https://github.com/ProkopHapala/FireCore/blob/f3423ab01e94a9021a7175250ac8f8f9f7005d5f/cpp/common_resources/cl/relax_multi_mini.cl#L1258) versus 2.10 seconds for texture based solution ( see [inline float4 fe3d_pbc_comb_tex_()](https://github.com/ProkopHapala/FireCore/blob/f3423ab01e94a9021a7175250ac8f8f9f7005d5f/cpp/common_resources/cl/relax_multi_mini.cl#L1634) and [__kernel void sampleGrid_tex()](https://github.com/ProkopHapala/FireCore/blob/f3423ab01e94a9021a7175250ac8f8f9f7005d5f/cpp/common_resources/cl/relax_multi_mini.cl#L2001) ), on my desktop GPU nVidia RTX 3090.

**Trying to rationalize this result and put into perspective:**  RTX 3090 has 10496 cores, so roughly speaking, we can say that each core can do one particle update (which is dominated by the interpolation) in 1-2 microsecond (2000-4000 clock cycles, at frequency 1.7GHz). That is significantly longer than on CPU, where one thread can do it in ~250 clock cycles (at 4-5GHz). But CPU has SIMD AVX512, better cache, and it is less memory bound.

--- 

PS: Maybe I should put this rather in some other issue, where we discuss tricubic interpolation, but I found it relevant to the dissusion abouve about  `read_imagef`

**Comment by NikoOinonen** [comment]
@ProkopHapala We had this other issue where we discussed the cubic interpolation before: #144.

**Comment by mondracek** [comment]
Let me admit that during that as I was implementing the `reverse` branch, I began to understand why things had been implemented before the way they had. There had in fact not been that many inconsistencies in the array indexing after all, the arrays were mostly `C[iz,iy,ix,i]`. What was confusing was that the vector variables like `ndim` that were supposed to contain grid dimensions sometimes contained them in the standard order `(nx, ny, nz)`, other times as the value returned by `grid.shape`, that is, `(nz, ny, nx)` or `(nz, ny, nx, 3)`. Moreover, there are places in the code where the `[iz, iy, ix]` indexing is natural, namely
1. reading and writing the XSF files. The arrangement of the data in XSF is essentially `Fortran[ix,i y, iz]`, which can be turned into `C[iz, iy, ix]` just by a formal redefinition of the array, without physical rearrangement of the data in memory. But we were not taking full advantage of this anyway; the `loadXSF` function always made a copy of the array it had read so as to really guarantee a particular memory layout.
2. More importantly, however, the reversed (`[iz, iy, ix]`) order of indices was really natural for the `plotImages` function and several similar functions (`plotVecFieldRG`, `plotDistortions`, `plotArrows`) from `PPPlot.py`, which plot 2D constant-height slices of 3D data. I was essentially forced to add lines that transpose the array(s) at the beginning of `plotImages` and other such functions; see [link to plotImages](https://github.com/Probe-Particle/ppafm/blob/681b393cc006cba0ec7012b02bed66db7b8ff5ec/ppafm/PPPlot.py#L89C5-L93C40) This also complicated the way one can specify that all slices are meant to be plotted, so as a bonus, I added the feature that the `slices` can be left undefined and that means the whole range of slices (`0 <= iz < nz`) will be plotted.

If I were fully aware of the issue with `PPPlot.py` functions before, I might not have decided to go on with this switching to the `C[ix, iy, iz] ` ordering after all, but I did not want to give up when I was already in the process of doing it after such a long discussion.

**Comment by NikoOinonen** [comment]
> However, some automatic tests failed as I uploaded the branch to GitHub. I may need some help with analyzing the problematic results of the automatic tests.

@mondracek Note that you can also run the tests locally without pushing anything. If you have `pytest` installed (get separately from pip or install ppafm with the `dev` option), then you can just run `pytest` at the root of the repo to run all of the tests, or run tests in a single file with `pytest path/to/test.py`. This way you can debug locally more easily.

**Comment by mondracek** [comment]
Mostly for @ProkopHapala, perhaps @NikoOinonen may know something about this issue:
Going back to the duplicate definitions in `ProbeParticle.cpp` and `GridUtils.cpp`, I have spotted a potentially more serious problem than the double definitions of the `setGridN()` and `setGridCell()` functions. The global variable (class instance) **`gridShape`** is declared in both `ProbeParticle.cpp` and `GridUtils.cpp`. I really do not know what happens when we use both modules together, as we actually do in `ppafm`. The way I understood the C language, I thought this should result in an error during the linking stage, but obviously, it does not. So the question is, will the `gridShape` be allocated in memory only once (as we probably want), or will it be allocated twice, for functions from `ProbeParticle.cpp` to use one of the instances and functions from `GridUtils.cpp` use the other? The latter could easily result in a disaster, which apparently has not happened yet, but that may be either because the linker ensures in fact only one instance of `gridShape` or simply because so far we have not been using that much those functions from `GridUtils.cpp` that would rely on `gridShape`.

**Comment by mondracek** [comment]
* I have created a new branch `cleanup`, based on the branch `reverse`. I am planning to implement further changes concerning this issue in this `cleanup` branch. 
* Besides the thing with index order taken from the branch `reverse`, I have addressed the issue with `setFF` and similar functions, also discussed in this thread (almost 2 years ago). Essentially, I have removed the functions `setFF` and `setFF_shape` from `ppafm/core.py` and replaced the function `prepareArrays`in `ppafm/HighLevel.py` by a new function, which I also call `setFF`. This new `setFF` integrates all the functionalities of the old `prepareArrays`, `setFF_shape`, and `setFF` and it is now used throughout the code instead of the old functions.
* I have created a pull request #347 for the `cleanup` branch. I have tested it on a better part of the examples in the `examples` folder and it passes the automatic tests. Please consider whether we should merge it into `main` now or wait until I implement some more changes.

**Comment by ProkopHapala** [comment]
> Essentially, I have removed the functions setFF and setFF_shape from ppafm/core.py and replaced the function prepareArraysin ppafm/HighLevel.py by a new function, which I also call setFF. This new setFF integrates all the functionalities of the old prepareArrays, setFF_shape, and setFF and it is now used throughout the code instead of the old functions.

aha, is it not too much overhead? I mean, sometimes it may be convenient just switch some memory pointer, rather than setup all the dimensions or even allocate new arrays. 

I like the original desing where we have high-level functions doing user-friendly high-level stuff, which call low-level function allowing fine-grain control to power-user

like this 

```python
def prepareArrays( FF, Vpot ):
    if ( FF is None ):
        gridN = PPU.params['gridN']
        FF    = np.zeros( (gridN[2],gridN[1],gridN[0],3)    )
    else:
        PPU.params['gridN'] = np.shape( FF )
    core.setFF_Fpointer( FF )
    if ( Vpot ):
        V = np.zeros( (gridN[2],gridN[1],gridN[0])    )
        core.setFF_Epointer( V )
    else:
        V=None
    #core.setFF( gridF=FF, gridE=V )
    return FF, V 
```

I would not delete `setFF_Fpointer()` and `setFF_shape()` and `setFF()`

**Comment by ProkopHapala** [comment]
Thanks for explaining Martin. If the low level functions are still available than it is good. I understand the original version which emerged historically required refactoring.
